albatross/core/fit_model.h

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (C) 2018 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef ALBATROSS_CORE_FIT_H
+#define ALBATROSS_CORE_FIT_H
+
+namespace albatross {
+
+template <typename ModelType, typename Fit>
+class FitModel {
+
+  template <typename X, typename Y, typename Z>
+  friend class Prediction;
+
+ public:
+
+  static_assert(std::is_move_constructible<Fit>::value,
+                "Fit type must be move constructible to avoid unexpected copying.");
+
+  FitModel(const ModelType &model,
+           const Fit &&fit)
+      : model_(model), fit_(std::move(fit)) {}
+
+ public:
+
+  template <typename PredictFeatureType>
+  Prediction<ModelType, PredictFeatureType, Fit>
+  get_prediction(const std::vector<PredictFeatureType> &features) const {
+    return Prediction<ModelType, PredictFeatureType, Fit>(*this, features);
+  }
+
+  const ModelType &model_;
```

Comment:
For now I'm storing a reference to the model.  So far though, these models are pretty lightweight, I could see it being "OK" to copy them each time you call fit.  On the other hand the `Fit` could be pretty large so we for sure don't want to copy it, hence the strict enforcement of `is_move_constructible` above.

Reply:
Like we discussed, this is fine as long as the object `model_` references remains valid for as long as you're using the `FitModel` objects. This is highly dependent on how Albatross is used, and is probably not something that easily enforced from inside of Albatross.

When we talked I, for some reason, was thinking that we didn't want to use any dynamic memory in Albatross. In hindsight I don't think that's strictly the case, it's not being used in an embedded context. This might be a good place to consider using `std::shared_ptr` and `std::weak_ptr` to more gracefully handle making sure pointers/references are valid before use.

Reply:
For now, what if we just made copies of `model_`? I'll look into what switching to smart pointers would involve.